File: src/lib/dtd/schemas/npc-templates.ts
/**
 * Zod schema for NPC template data — mirrors pipeline/models/npc_templates.py.
 */

import { z } from "zod";
import { SkillRef } from "./common.js";

// ---------------------------------------------------------------------------
// NpcCharacteristics (strict)
// ---------------------------------------------------------------------------

export const NpcCharacteristics = z
	.object({
		strength: z.number().int(),
		dexterity: z.number().int(),
		constitution: z.number().int(),
		charisma: z.number().int(),
		fellowship: z.number().int(),
		composure: z.number().int(),
		intelligence: z.number().int(),
		wisdom: z.number().int(),
		willpower: z.number().int(),
	})
	.strict();
export type NpcCharacteristics = z.infer<typeof NpcCharacteristics>;

// ---------------------------------------------------------------------------
// NpcArmor (lenient)
// ---------------------------------------------------------------------------

export const NpcArmor = z
	.object({
		name: z.string(),
		ap: z.number().int().nullish(),
		locations: z.array(z.string()).nullish(),
	})
	.passthrough();
export type NpcArmor = z.infer<typeof NpcArmor>;

// ---------------------------------------------------------------------------
// NpcWeapon (lenient)
// ---------------------------------------------------------------------------

export const NpcWeapon = z
	.object({
		name: z.string(),
		type: z.string(),
		damage: z.string(),
		damageType: z.string(),
		pen: z.number().int(),
		special: z.union([z.string(), z.array(z.string())]).default(""),
		range: z.number().int().nullish(),
		rof: z.string().nullish(),
		clip: z.number().int().nullish(),
		reload: z.string().nullish(),
	})
	.passthrough();
export type NpcWeapon = z.infer<typeof NpcWeapon>;

// ---------------------------------------------------------------------------
// NpcTraitRef (lenient)
// ---------------------------------------------------------------------------

export const NpcTraitRef = z
	.object({
		id: z.string(),
		param: z.union([z.string(), z.number().int()]).nullish(),
	})
	.passthrough();
export type NpcTraitRef = z.infer<typeof NpcTraitRef>;

// ---------------------------------------------------------------------------
// NpcAbility (lenient)
// ---------------------------------------------------------------------------

export const NpcAbility = z
	.object({
		name: z.string(),
		description: z.string(),
	})
	.passthrough();
export type NpcAbility = z.infer<typeof NpcAbility>;

// ---------------------------------------------------------------------------
// NpcTemplate (lenient)
// ---------------------------------------------------------------------------

export const NpcTemplate = z
	.object({
		id: z.string(),
		name: z.string(),
		category: z.string(),
		level: z.number().int(),
		size: z.number().int(),
		speed: z.number().int(),
		characteristics: NpcCharacteristics,
		skills: z.array(SkillRef),
		feats: z.array(z.string()),
		traits: z.array(z.union([z.string(), NpcTraitRef])),
		armor: z.array(NpcArmor),
		weapons: z.array(NpcWeapon),
		abilities: z.array(z.union([z.string(), NpcAbility])),
		gear: z.array(z.string()),
	})
	.passthrough();
export type NpcTemplate = z.infer<typeof NpcTemplate>;

// ---------------------------------------------------------------------------
// Top-level file schema (bare array)
// ---------------------------------------------------------------------------

export const NpcTemplatesFile = z.array(NpcTemplate);
export type NpcTemplatesFile = z.infer<typeof NpcTemplatesFile>;

File: pipeline/models/npc_templates.py
"""Pydantic model for npc-templates.json."""

from __future__ import annotations

from .common import LenientModel, SkillRef, StrictModel


class NpcCharacteristics(StrictModel):
    """The 9 characteristics for an NPC."""

    strength: int
    dexterity: int
    constitution: int
    charisma: int
    fellowship: int
    composure: int
    intelligence: int
    wisdom: int
    willpower: int


class NpcArmor(LenientModel):
    """Armor equipped by an NPC."""

    name: str
    ap: int | None = None
    locations: list[str] | None = None


class NpcWeapon(LenientModel):
    """An inline weapon definition for an NPC."""

    name: str
    type: str  # "melee" or "ranged"
    damage: str
    damageType: str
    pen: int
    special: str | list[str] = ""
    # Ranged-only optional fields
    range: int | None = None
    rof: str | None = None
    clip: int | None = None
    reload: str | None = None


class NpcTraitRef(LenientModel):
    """A trait reference — can be a string or a dict with id + param."""

    id: str
    param: str | int | None = None


class NpcAbility(LenientModel):
    """A special ability — can be a string or a dict with name + description."""

    name: str
    description: str


class NpcTemplate(LenientModel):
    """A single NPC template. Note: uses kebab-case IDs unlike most other files."""

    id: str
    name: str
    category: str
    level: int
    size: int
    speed: int
    characteristics: NpcCharacteristics
    skills: list[SkillRef]
    feats: list[str]
    traits: list[str | NpcTraitRef]
    armor: list[NpcArmor]
    weapons: list[NpcWeapon]
    abilities: list[str | NpcAbility]
    gear: list[str]
